odl/deform/linear_deformation.py

```diff
@@ -0,0 +1,192 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Shape-based reconstruction using optimal information transportation.
+
+The Fisher-Rao metric is used in regularization term. And L2 data matching
+term is used in fitting term.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+from odl.operator import Operator
+from odl.discr import DiscreteLp, Divergence
+from odl.space import ProductSpace
+
+__all__ = ('LinearDeformation', 'MassPreservingLinearDeformation')
+
+
+class LinearDeformation(Operator):
+    """Linearized deformation."""
+    def __init__(self, offsets=None):
+        """Initialize a linear deformation
+
+        This has the form
+
+        Id + offsets
+
+        Parameters
+        ----------
+        vector_field : `ProductSpace` of `DiscreteLp` spaces
+            The points to calculate the interpolation in.
+
+        Examples
+        --------
+        Identity deformation:
+
+        >>> import odl
+        >>> space = odl.uniform_discr(0, 1, 5)
+        >>> pspace = ProductSpace(space, space.ndim)
+        >>> vector_field = pspace.element(space.points().T)
+        >>> deformation = LinearDeformation(vector_field)
+        """
+
+        assert isinstance(offsets.space, ProductSpace)
+        assert offsets.space.is_power_space
+        assert isinstance(offsets.space[0], DiscreteLp)
+
+        ndim = offsets.space.size
+        space = offsets.space[0]
+
+        interp_points = space.points().T
+        for i in range(ndim):
+            interp_points[i] += offsets[i].ntuple.asarray()
+
+        self.interp_points = interp_points
+        self.offsets = offsets
+
+        Operator.__init__(self,
+                          domain=space,
+                          range=space,
+                          linear=True)
+
+    def _call(self, x):
+        return x.interpolation(self.interp_points, bounds_check=False)
```

Comment:
Perhaps allow using other interpolations than the space interpolation here. I'm not 100% sure.